packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,415 @@
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    new instanceDefinition.UserCtor();
+
+    patchAttributesDuringUpgrade(instance, registeredDefinition, instanceDefinition);
+}
+
+function getDefinitionForConstructor(constructor: CustomElementConstructor): Definition {
+    if (!isFunction(constructor) || !isObject(constructor.prototype)) {
+        throw new TypeError('The referenced constructor is not a constructor.');
+    }
+    let definition = definitionForConstructor.get(constructor);
+    if (isUndefined(definition)) {
+        definition = createDefinitionRecord(constructor);
+        definitionForConstructor.set(constructor, definition);
+    }
+    return definition;
+}
+
+export function patchCustomElementRegistry() {
+    const { customElements: nativeRegistry } = window;
+    const { define: nativeDefine, whenDefined: nativeWhenDefined, get: nativeGet } = nativeRegistry;
+
+    // patch for the global registry define mechanism
+    CustomElementRegistry.prototype.define = function define(
+        this: CustomElementRegistry,
+        tagName: string,
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions
+    ): void {
+        if (options && options.extends) {
+            // TODO [#2877]: should we support `extends`?
+            throw new DOMException(
+                'NotSupportedError: "extends" key in customElements.define() options is not supported.'
+            );
+        }
+        nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.
+        if (!isUndefined(globalDefinitionsByTag.get(tagName))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': the name "${tagName}" has already been used with this registry`
+            );
+        }
+        const definition = getDefinitionForConstructor(constructor);
+        if (!isUndefined(globalDefinitionsByClass.get(constructor))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': this constructor has already been used with this registry`
+            );
+        }
+        globalDefinitionsByTag.set(tagName, definition);
+        globalDefinitionsByClass.set(constructor, definition);
+        let PivotCtor = pivotCtorByTag.get(tagName);
+        if (isUndefined(PivotCtor)) {
+            PivotCtor = createPivotingClass(tagName, definition);
+            // Register a pivoting class which will handle global registry initializations
+            nativeDefine.call(nativeRegistry, tagName, PivotCtor);
+        }
```

Comment:
I refactored it a bit because of https://github.com/salesforce/lwc/pull/2724#discussion_r897049884, but I'm not sure it's worth extracting the similar logic into a shared function. They have slightly different ordering in how they do things. Let me know what you think of the new version.